supporter-product-data/src/main/scala/com/gu/lambdas/ProcessSupporterRatePlanItemLambda.scala

```
@@ -0,0 +1,132 @@
+  }
+}
+
+class ContributionProcessor(stage: Stage, config: ZuoraQuerierConfig) {
+  val contributionIds = ContributionIds.forStage(stage)
+  lazy val zuoraService = new ZuoraSubscriptionService(config, configurableFutureRunner(60.seconds))
+
+  def isRecurringContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    contributionIds.contains(supporterRatePlanItem.productRatePlanId)
+
+  def addAmountIfContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    if (isRecurringContribution(supporterRatePlanItem)) {
+      SafeLogger.info(s"Supporter ${supporterRatePlanItem.identityId} is a recurring contributor")
+      fetchContributionAmountFromZuora(supporterRatePlanItem)
+    } else
+      Future.successful(supporterRatePlanItem)
+
+  def fetchContributionAmountFromZuora(supporterRatePlanItem: SupporterRatePlanItem) =
+    zuoraService
+      .getSubscription(supporterRatePlanItem.subscriptionName)
+      .map { sub =>
+        SafeLogger.info(
+          s"Contribution amount for supporter ${supporterRatePlanItem.identityId} is ${sub.contributionAmount}",
+        )
+        supporterRatePlanItem.copy(contributionAmount = sub.contributionAmount)
+      }
+
+}
+
+class SupporterRatePlanItemProcessor(
+    dynamoService: SupporterDataDynamoService,
+    alarmService: AlarmService,
+    contributionProcessor: ContributionProcessor,
+) {
+
+  def processItem(supporterRatePlanItem: SupporterRatePlanItem) =
+    contributionProcessor
+      .addAmountIfContribution(supporterRatePlanItem)
+      .flatMap(writeItemToDynamo)
+
+  def writeItemToDynamo(supporterRatePlanItem: SupporterRatePlanItem) = {
+    SafeLogger.info(s"Attempting to write item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
+    try {
+      dynamoService.writeItem(supporterRatePlanItem).map { _ =>
+        SafeLogger.info(s"Successfully wrote item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
+        ()
+      }
+    } catch {
```

this won't do much as writeItem returns a Future - we need to be looking out for a Failed future which I think will happen at the handlerFuture level.  Ideally we can catch all exceptions at the lowest level, and deal in Futures and Trys everywhere else.